### PR TITLE
Support better formatting for curried arrow functions

### LINKED ISSUE #9985: Bad Formatting: Curried Function definitions with lots of parameters
This example is in JavaScript, but the same problem also happens with TypeScript

**Prettier 2.2.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEYCuAnPATwBU4NMBeAHSkzswAoBDPAcxwFsEYBGASkoA+GvUYt2XWACYBFYbXrM2nbgGZZ80UoncALBpGLxK2AFYDCutpMwAbBa3HJMAOwOjy5wA53Vp9wBOXzFPbh4ABmDrZx5+IUM-UNgeGXjLTCoYBmAEjJhMgDMICCRMf2SAGlzMzIAjFlLymCkq9MyAXz4AbhoQCpAIAAcYAEt0ZFAWPAgAdwAFFgQ0ZBAmABsZpkJl-tq8JjAAazgYAGVBg5GoVmQYPBw4frgOWrgAEzf3gBkma5wmVhwABiEDwHCYMFG1xWTBwMAgfRAAAsYBw1gB1JEjeBoC5gOCnJbYkYAN2xhBWYDQOxAVzQcDwMDm+1Y4OQBXW9P6ACs0AAPABC+yOJ1OTC4XyucHZnMeIF5fNOV1YazgAEUcBB4DK1lyQBc8PS8Ct6q81ojBngrjB0SM3jAkcgvOF+paIPT0ftBitLWQGSTpf0AI6a+DMobLFBMNAAWigcHe70ReDgIZGKeZALZSA5url9I4I1u93zytVGq10pzsv6MCYtVt9sdSF0tf2IzWyoAwhAONmQGRTIicPTiPXI7m9SSHgBJKCfWCnMBW4YAQXnpxghFVOvp7XaQA)

**Input:**

```jsx
const curryTest =
    (argument1) =>
    (argument2) =>
    (argument3) =>
    (argument4) =>
    (argument5) =>
    (argument6) =>
    (argument7) =>
    (argument8) =>
    (argument9) =>
    (argument10) =>
    (argument11) =>
    (argument12) =>
    	({
    		foo: argument1,
    		bar: argument2,
    	});
```

**Output:**

```jsx
const curryTest = (argument1) => (argument2) => (argument3) => (argument4) => (
    argument5
) => (argument6) => (argument7) => (argument8) => (argument9) => (
    argument10
) => (argument11) => (argument12) => ({
    foo: argument1,
    bar: argument2,
});
```

**Expected behavior:**

Something closer to the input

**Notes**

  - I am writing code in this style because I am doing Applicative validation(parsing) of data structures that contain many fields
  - This is a different issue from #4061
  - This merged PR seems related: #1066

### PR DESCRIPTION
**Pull request by sosukesuzuki**
## Description

Fixes #9985

## Checklist

- [x] I’ve added tests to confirm my change works.
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: src/language-js/print/function.js
Comment (fisker) [inline src/language-js/print/function.js]: Should return `ture` in loop, when count >=3, but do you think better use `2`?

Comment (fisker) [reply src/language-js/print/function.js]: I think you misunderstood, I mean

```js
while () {
  if (arrowChainCount >= 2) return true
}
```

No need loop to the most inner one.

And I didn't mean `>= 3`.

Comment (sosukesuzuki) [reply src/language-js/print/function.js]: Ah I got it, thanks. I'll fix.
Btw which level do you think we should choose?

Comment (fisker) [reply src/language-js/print/function.js]: I think `2`,

`const a = (b) => (c) => 1`, this is `2`, right ?

Comment (sosukesuzuki) [reply src/language-js/print/function.js]: Yes. I changed to choose level `2` at 731a12b. The commit updates some test cases, please check. I personally think we can approve those updates.

**Comment by sosukesuzuki** [comment]
Tests are failed. I'll look into this later.

File: tests/js/arrows/__snapshots__/jsfmt.spec.js.snap
Comment (fisker) [inline tests/js/arrows/__snapshots__/jsfmt.spec.js.snap:692]: This seems not pretty.

Comment (fisker) [reply tests/js/arrows/__snapshots__/jsfmt.spec.js.snap:692]: Maybe no break before first level parameters more reasonabe.

Comment (sosukesuzuki) [reply tests/js/arrows/__snapshots__/jsfmt.spec.js.snap:692]: what do you mean? Should be like below?:
```js
fooooooooooooooooooooooooooooooooooooooooooooooooooo((action) =>
  (next) => dispatch(action)
);
```

Comment (fisker) [reply tests/js/arrows/__snapshots__/jsfmt.spec.js.snap:692]: Yes, I mean this, but I'm not sure. Better idea?

Comment (thorn0) [reply tests/js/arrows/__snapshots__/jsfmt.spec.js.snap:692]: I'd prefer not breaking the chain if it fits on one line:
```js
fooooooooooooooooooooooooooooooooooooooooooooooooooo(
  (action) => (next) => dispatch(action)
);
```
And if the chain is long enough, then:
```js
fooooooooooooooooooooooooooooooooooooooooooooooooooo(
  (action) =>
    (next) =>
    (next) =>
    (next) =>
    (next) =>
    (next) =>
    (next) => 
    dispatch(action)
);
```
This is similar to how Prettier formats binary expressions in call arguments.

Comment (fisker) [reply tests/js/arrows/__snapshots__/jsfmt.spec.js.snap:692]: In that case, assign print like this?

```js
const foo = (action) =>
  (next) =>
  (next) =>
  (next) =>
  (next) =>
  (next) =>
  (next) => 
  dispatch(action)
```

Comment (thorn0) [reply tests/js/arrows/__snapshots__/jsfmt.spec.js.snap:692]: Assignments are complicated. To make the chain look uniform and to avoid dealing with edge cases (#2482), it's better to break after `=`.

```js
const foo =
  (action) =>
  (next) =>
  (next) =>
  (next) =>
  (next) =>
  (next) =>
  (next) => 
  dispatch(action)
```

Comment (sosukesuzuki) [reply tests/js/arrows/__snapshots__/jsfmt.spec.js.snap:692]: Thank you. I agree with thorn0's suggestions.

File: src/language-js/utils.js
Comment (brodybits) [inline src/language-js/utils.js:1345]:
```suggestion
  return (
    node &&
    node.type === "ArrowFunctionExpression" &&
```

File: src/language-js/utils.js
Comment (brodybits) [inline src/language-js/utils.js:1342]: Hi sorry to bother you with another question: do we know if the `node` object reference could ever be missing here and can we think of a way we could test this?

I proposed the simplification here in response to a codecov warning but am now wondering if someone else might think of removing this condition at some point in the future? Or am I getting too fussy here?

Comment (fisker) [reply src/language-js/utils.js:1342]: Yes, it could be `call(, foo)`, it's not valid code, but babel can parse it.

File: tests/js/arrows/__snapshots__/jsfmt.spec.js.snap
Comment (fisker) [inline tests/js/arrows/__snapshots__/jsfmt.spec.js.snap:2733]: Can you add indention for this? I think this is `CallExpression.callee`?

File: src/language-js/utils.js
Comment (brodybits) [reply src/language-js/utils.js:1342]: If I remove this condition on the `node` object reference, all tests still pass in my work area. When I tried it with `call(, foo)`, it leads to an exception regardless of this condition on the `node` object reference. When I tried merging with PR #10023 and removing this condition, still no test failures. So I am not really convinced at this point.

File: tests/js/arrows/__snapshots__/jsfmt.spec.js.snap
Comment (sosukesuzuki) [reply tests/js/arrows/__snapshots__/jsfmt.spec.js.snap:2733]: like below?
```js
(
  argument1 =>
    argument2 =>
    argument3 =>
    argument4 =>
    argument5 =>
    argument6 =>
    argument7 =>
    argument8 =>
    argument9 =>
    argument10 =>
    argument11 =>
    argument12 =>
    3
)(3);
```

Comment (fisker) [reply tests/js/arrows/__snapshots__/jsfmt.spec.js.snap:2733]: Yes, WDYT?

Comment (sosukesuzuki) [reply tests/js/arrows/__snapshots__/jsfmt.spec.js.snap:2733]: It was intended to be unified with the format for binary expressions. But I have no strong opinion.

**Prettier pr-9992**
[Playground link](https://deploy-preview-9992--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEAKAOlABF1AzCQoiASiwF4A+THfYosqm3A+0i67Ft9pruto061W9Ic1FtMJVCQDcmTBn6SGWANTMBYjVtWEymlT0N6Tu44Is59RabIVQQAGhAQADjACW0AM7JQAEMAJ2CIAHcABRCEfxRAgBtwwIBPf1cAI2DAsABrOBgAZXccrygAc2QYYIBXOFc4AFsMuAATVraAGUCKmsDyuAAxCGDGwJhvCuQQQJqYCBcQAAsYRoSAdSWveF8SsDhC2O2vADdtlOmwX3SQMt84YJhI7PKx5DxE+9cAK18ADwAQtk8gVCoFGnBOmU4O9PvUQL8-oUyuUEnAAIo1CDwWEJL4gErBe7BaYZQItBKLdzBMowdZeVowJbIAAcAAZXNSIPd1tl3NNqXBiScYa4AI5Y+DPDxxGa+AC0UDgbTai2CcAlXnVz36byQHzx8PujS8VVqRpRaMx2Jh+rhrhg5PpjOZSAATA7sl4EiiAMIQRp6kBCgCsixq9wAKuS4gb8Sc6gBJKAdWCFMA0zwAQRThRgKTRuPuAF9i0A)
```sh
--parser babel
```

**Input:**
```jsx
(
  (fooooooo) =>
  (fooooooo) =>
  (fooooooo) =>
  (fooooooo) =>
  (fooooooo) =>
  (fooooooo) =>
  fooooooooo
)();

(
  (fooooooo) +
  (fooooooo) +
  (fooooooo) +
  (fooooooo) +
  (fooooooo) +
  (fooooooo) +
  fooooooooo
)();

```

**Output:**
```jsx
(
  (fooooooo) =>
  (fooooooo) =>
  (fooooooo) =>
  (fooooooo) =>
  (fooooooo) =>
  (fooooooo) =>
  fooooooooo
)();

(
  fooooooo +
  fooooooo +
  fooooooo +
  fooooooo +
  fooooooo +
  fooooooo +
  fooooooooo
)();

```

**Comment by sosukesuzuki** [comment]
Can anyone review?

File: src/language-js/utils.js
Comment (fisker) [inline src/language-js/utils.js:1337]: I know this is silly, but can we add `import()` too?

Comment (fisker) [reply src/language-js/utils.js:1337]: **Prettier pr-9992**
[Playground link](https://deploy-preview-9992--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEBLAtgBwgJxgCgAJ8BDHAcwFd0EYBGASkIF4A+AHSkO+LKptgAmJm049eFarQDMIjlx6lJAmABY5YxXymwArBoXcl-WgDYD44zpgB2C1uW0AHPaPaVATlcSTsOgAZvKxU6RhZ5S3daOmFwwmBNcUgoAGcYQgAzCAgWQnYQLIh8gG5Enhw4GEocLkLCAGo8kAAjMhKygF8GYpAAGhAITBhUaBTkUDIcCAB3AAUyBDGUEgAbaZIATzH+5pwSMABrSoBlTH3UKHJkGBxKOH64dGa4ABMX14AZEkvKEnI4ABiuHQJBgw0uyBAJEoMAgfRAAAsYOgVgB1BGoeApM5gODHRaY1AAN0xG0hYBS2zQqTgeFme3IIOQGVWKXuIAAVikAB4AIT2hxOJBoHwucGZrPZXO5xwu5BWcAAipQIPAJSs2f0zjg2ThIa1nit4ZgcBcYKjUC8YAjkE5-FqpmzUXtMJDVustvCAI4q+D0wZLKEpAC0UDgr1e8IqPtQFXpfyZSBZGvZbPQqGut1TcoVytV4qTkv6MBIzQtVptSEExb2qBWcoAwhB0ImQHAUrp4ZQ2QAVUtLZOakBEu4ASSg71gxzApqGAEEJ8cYBsFeq2R0OkA)
```sh
--parser babel
--prose-wrap always
```

**Input:**
```jsx
import( (argument1) =>
    (argument2) =>
    (argument3) =>
    (argument4) =>
    (argument5) =>
    (argument6) =>
    (argument7) =>
    (argument8) =>
    (argument9) =>
    (argument10) =>
    (argument11) =>
    (argument12) => {
      const foo = "foo";
      return foo + "bar";
    });
```

**Output:**
```jsx
import(
  (argument1) =>
    (argument2) =>
    (argument3) =>
    (argument4) =>
    (argument5) =>
    (argument6) =>
    (argument7) =>
    (argument8) =>
    (argument9) =>
    (argument10) =>
    (argument11) =>
    (argument12) => {
      const foo = "foo";
      return foo + "bar";
    }
);

```

**Prettier pr-9992**
[Playground link](https://deploy-preview-9992--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEBLAtgBwgJxgCgAJ8BDHAcwFd0EYBGASkIF4A+AHSkO+LKptgAmJm049eFarQDMIjlx6lJAmABY5YxXymwArBoXcl-WgDYD44zpgB2C1uW0AHPaPaVATlcSTsOgAZvKxU6RhZ5S3daOmFwwmBNcUgoAGcYQgAzCAgWQnYQLIh8gG5Enhw4GEocLkLCAGo8kAAjMhKygF8GYpAAGhAITBhUaBTkUDIcCAB3AAUyBDGUEgAbaZIATzH+5pwSMABrSoBlTH3UKHJkGBxKOH64dGa4ABMX14AZEkvKEnI4ABiuHQJBgw0uyBAJEoMAgfRAAAsYOgVgB1BGoeApM5gODHRaY1AAN0xG0hYBS2zQqTgeFme3IIOQGVWKXuIAAVikAB4AIT2hxOJBoHwucGZrPZXO5xwu5BWcAAipQIPAJSs2f0zjg2ThITQcKgNiQEfDMIbYKjUC8YKakE5-FqpmzUXtMJDVustvCAI4q+D0wZLKEpAC0UDgr1e8IqftQFXpfyZSBZGvZbPQqGut3TcoVytV4pTkv6MBIzStNrtglLe1QKzlAGEIOhkyA4CldPDKGyACrlpapzUgIl3ACSUHesGOYENQwAgpPjjANgr1WyOh0gA)
```sh
--parser meriyah
--prose-wrap always
```

**Input:**
```meriyah
import( (argument1) =>
    (argument2) =>
    (argument3) =>
    (argument4) =>
    (argument5) =>
    (argument6) =>
    (argument7) =>
    (argument8) =>
    (argument9) =>
    (argument10) =>
    (argument11) =>
    (argument12) => {
      const foo = "foo";
      return foo + "bar";
    });
```

**Output:**
```meriyah
import(
  (argument1) =>
  (argument2) =>
  (argument3) =>
  (argument4) =>
  (argument5) =>
  (argument6) =>
  (argument7) =>
  (argument8) =>
  (argument9) =>
  (argument10) =>
  (argument11) =>
  (argument12) => {
    const foo = "foo";
    return foo + "bar";
  }
);

```

Comment (sosukesuzuki) [reply src/language-js/utils.js:1337]: We cannot add `ImportExpression` to `isCallLikeExpression` because `import(...)` isn't like call expression. But I changed to format import and curried functions in the same way as call expressions ( 2de96ba ).

File: src/language-js/print/function.js
Comment (fisker) [inline src/language-js/print/function.js:214]: Why not add  to  `isCallLikeExpression`

File: src/language-js/utils.js
Comment (fisker) [reply src/language-js/utils.js:1337]: Sorry, didn't see this, I think we should, we already treat it like `CallExpression` everywhere.

File: src/language-js/print/function.js
Comment (sosukesuzuki) [reply src/language-js/print/function.js:214]: Because AST form is different between `ImportExpression` and `CallExpression` and `isCallLikeExpression` is used for not only printing curried functions. For example, we use `node.arguments` after checking with `isCallLikeExpression` (https://github.com/prettier/prettier/pull/9992/files#diff-112b7d78917580aa0cc6cc3ac0e641c6a9541c0ed4de574f05dba2ae51480ccaR525-R526)

File: src/language-js/utils.js
Comment (sosukesuzuki) [reply src/language-js/utils.js:1337]: What do you think about https://github.com/prettier/prettier/pull/9992#discussion_r558318092 ?

Comment (fisker) [reply src/language-js/utils.js:1337]: This why I added https://github.com/prettier/prettier/blob/2f3fc241f8cb1867a0a3073ceac9d662e4b25fab/src/language-js/utils.js#L1210-L1235, should not use `node.arguments`.

Comment (sosukesuzuki) [reply src/language-js/utils.js:1337]: Okay, I'll fix.

Comment (sosukesuzuki) [reply src/language-js/utils.js]: done 6fb35ae

**Review by fisker — APPROVED** [review]

**Comment by thorn0** [comment]
I need some more time to look into this. Sorry for the delay.

**Comment by thorn0** [comment]
I'm afraid this change might be too disruptive. Before merging we need to test it on a couple of codebases that heavily use curried arrow functions. Anybody knows of such projects we can use for testing?

**Comment by thorn0** [comment]
Looks very good overall. I have comments on two things.

(1) What's the story behind this difference?

**Prettier pr-9992**
[Playground link](https://deploy-preview-9992--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuc0DOMAEYCuAnPATwBU4NMBeAHSkzswAoBDPAcxwFsEYBGASkoA+GvUYt2XWACYBFQY2AjR9AGYQISTOM7ceAGiXLMAIxabtkmFIO16AXz4BuGjUhRyuAiTJZqtusxsOrD8QoZiQZYyYf4REtwAzLLyDIqxomoaWpG6NkZ0pnjmOdJ59k4uUCB6IBAADjAAlujIoCx4EADuAAosCGjIIEwANp1MhAM1xnhMYADWcDAAynWzjVCsyDB4OHA1cBzGcAAmxycAMkwbOEyscABiEHgcTDBNG4NMODAQ1SAAFjAOMMAOr-RrwNCrMBwJb9CGNABuEMIgzAaEmIHWaDgeBg3RmrBeyBUIxxNQAVmgAB4AIRm80WSyYXHO6zgJLJexAVOpS3WrGGcAAijgIPBOcNySBVngcXhBqYjsM-nU8OsYCDGscYP9kAAOAAMNTVEBxIJmdUGarIuMRHJqAEcxfACfUBigmGgALRQOAnE5-PBwZ2NYME27EpCkqXcnEcRpbHZxgVC0XijnRrk1GBMYxanV6pDWEDbJiNYYCgDCEA4UZAZAArH8cDjiHmPTHpYjdgBJKBnWBLMDqhoAQQHSxghCFkpxdjsQA)
```sh
--parser babel
```

**Input:**
```jsx
const curryTest =
    (argument1) =>
    (argument2) => ({
        foo: argument1,
        bar: argument2,
    });

const curryTest =
    (argument1) =>
    (argument2) =>
    (argument3) => ({
        foo: argument1,
        bar: argument2,
    });


```

**Output:**
```jsx
const curryTest =
  (argument1) =>
  (argument2) => ({
    foo: argument1,
    bar: argument2,
  });

const curryTest =
  (argument1) =>
  (argument2) =>
  (argument3) =>
  ({
    foo: argument1,
    bar: argument2,
  });

```

(2) After playing with different examples, I decided that after all it's better to add another level of indentation to the body when the body is an expression.

```js
const curryTest =
  (argument1) =>
  (argument2) =>
  (argument3) =>
    argument1 + argument2 + argument3;

const foo =
  (action) =>
  (next) =>
  (next) =>
  (next) =>
  (next) =>
  (next) =>
  (next) =>
    ({
      foo: argument1,
      bar: argument2,
    });
```

Looks a bit better to me than the current output:

```js
const curryTest =
  (argument1) =>
  (argument2) =>
  (argument3) =>
  argument1 + argument2 + argument3;

const foo =
  (action) =>
  (next) =>
  (next) =>
  (next) =>
  (next) =>
  (next) =>
  (next) =>
  ({
    foo: argument1,
    bar: argument2,
  });
```

**Review by thorn0 — APPROVED** [review]
This case looks a bit controversial. Probably, it's better to keep it as in input. But I'm not sure, would be good to hear other opinions on this. Overall, LGTM.

**Prettier pr-9992**
[Playground link](https://deploy-preview-9992--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuc0DOMAEBDTBeAHSk0wAoAPASnwD4yBPavO0gLyZqJJPMwHpMABgB0ARgBMAZgAsAVgBsAdgAcATkwBqBpgBUmMNgA2YAK6Hs8AMoQAtnBgALAJZQA5m0rUBAWQsPhAAoAkgDcRCAANCAQAA4wTujIoNgATikQAO4BqQhoyCBGGdj0eVEARinYYADW9pYxVS6uyDApJnBRcDZlcAAmvX0AMthuJtiucABiECk2FvFu+dgmMBCRIA4wNoYA6s7waA1gcJa5TvEAbuf0+WBopSAuaHApMAGVrnPIAGZGz1EAKzQ5AAQpUanVsHZBi44D8-h0QEDyJYmoY4ABFEwQeDwwz-EANFLPFL5GD0GJwNBgFJOOLrGK02A7Jy9RzIZSCKKMiDPHaVGL5RlUl4XOFRACO2Pg71ieRQ2DQAFooHA+n11ik4FKnFr3uMvkhfvjEc8bE4Wm1TWjMdK4UaEVEYNgyiy2Q5kOInZUnIYmgBhWyGkBU2TrEzPAAqLvlxoJF3aQSgA1glhpdJgAEFk5Zyei8c8AL6FoA)
```sh
--parser typescript
```

**Input:**
```tsx
const a =
  (x) => (y) => (z) =>
    x / 0.123456789 + (y * calculateSomething(z)) / Math.PI;

```

**Output:**
```tsx
const a =
  (x) =>
  (y) =>
  (z) =>
    x / 0.123456789 + (y * calculateSomething(z)) / Math.PI;

```

**Review by thorn0 — CHANGES_REQUESTED** [review]
Sorry for taking back the approval, but turns out printing simple arrow chains ([my previous comment](https://github.com/prettier/prettier/pull/9992#pullrequestreview-584911801)) is intended behavior. See #1389

So let's put every `=>` on its own line only if the chain doesn't fit on one line.

Also your PR is going to fix https://github.com/prettier/prettier/issues/1389 and #6107. Please add tests from that issue.

**Comment by thorn0** [comment]
@sosukesuzuki I can't request your review because it's your PR, but please take a look. I tried to minimize the diff in https://github.com/prettier/prettier-regression-testing/issues/40 and only fix what was reported in #9985.

**Review by fisker — APPROVED** [review]
LGTM

**Review by sosukesuzuki — COMMENTED** [review]
I can't approve this PR on GitHub because I opened this.

But LGTM! Thank you!

**Comment by thorn0** [comment]
Oops. Sure, it's a corner case, but assignment chains and arrow chains don't play well together:

**Prettier pr-9992**
[Playground link](https://deploy-preview-9992--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEBDAvAI3QE3QCgB0oACE1AJwHMBXAWwRgHko4AaYsy2h2AFQDuEDqXLV6jPgAsKcOMQCUJdAD4SsgM7MAZtLgBBcbxgaAjACYAzMrXBO6uDBoVShEJlQAvNwG5iAXxA2EAgABxgAS2gNZFBKCggBAAVKBBiUVAAbAVQATxjgzApUMABrRwBlUJKIqCpkGAoadhA4Okw4HBxOgBlUOppUKjgAMQgKOlQYSLrkNBoYCCCQKRg6TIB1KQj4DWqwOAq0nYiANx3cubANApBajTgKGCTiqknkbSyH4IArDQAPABCxTKlVQDB6tTgHy+LT+-wqtSomTgAEUaBB4DDMt8QNUKA8KHMPB1MstQhRajANhEcDApMgAOwAVmCFIgDw2xVCcwpcEJp2hwQAjhj4C8wuk0BoALSsTqdZayUURWQvIbvJCfHEtB50CINJq6pEo9GY6Fa2HBGCoTA0ukMpDma3FCKZJEAYQgdE1rQ0zOWNAefFt6W1uNOzQAklBurAKmBKeF9LGKjBcijsQ9-P4gA)

```sh
--parser babel
--print-width 75
```

**Input:**

```jsx
a=b=d=(
  argumentOne,
  argumentTwo,
  argumentThree
) => restOfTheArguments123 => {
  return "baz";
}
```

**Output:**

```jsx
a =
  b =
  d =

      (argumentOne, argumentTwo, argumentThree) =>
      (restOfTheArguments123) => {
        return "baz";
      };

```

**Comment by onhate** [comment]
how do I rollback to the previous behavior?

**Comment by tony** [comment]
Am I correct in this not being toggleable / configurable?  I'm surprised by how much this indentation to swaths of our codebase 😆 

Personally I'm fine with the formatting (whatever's consistent) but I think something like this needs an "opt out" so people could upgrade prettier and delay the change.

In the README, it'd be good to get more details on if this is tweakable somehow.

**Comment by sosukesuzuki** [comment]
You(users) cannot rollback to prev behavior. But if you have actual code examples made harder to read by this feature, you can report it as a bug. And it'll help us to improve formatting heuristics.

**Comment by codeaid** [comment]
I upgraded to 2.3 earlier today and after running `prettier --check` it showed me around 20 files that are apparently incorrect despite being just fine before that. I suspect it has something to do with this PR. From what I gather there is no way to see what is actually wrong with them without running `--write` (convenient, not) but I eventually managed to find the problem:

```typescript
export const createCombinedReducer = (): Reducer<State> =>
    // Wrap root reducer to handle store reset actions
    (state, action) => {
        if (action.type === SystemAction.Reset) {
            return rootReducer(undefined, action);
        }

        return rootReducer(state, action);
    };
```

The reformatted version (100 chars wide) looks like this, which in my view is much uglier than before:

```typescript
export const createCombinedReducer =
    (): Reducer<State> =>
    // Wrap root reducer to handle store reset actions
    (state, action) => {
        if (action.type === SystemAction.Reset) {
            return rootReducer(undefined, action);
        }

        return rootReducer(state, action);
    };
```

If I remove the comment from line 3 it becomes something totally different again:

```typescript
export const createCombinedReducer = (): Reducer<State> => (state, action) => {
    if (action.type === SystemAction.Reset) {
        return rootReducer(undefined, action);
    }

    return rootReducer(state, action);
};
```

Is there a flag to disable moving `(): Reducer<State> =>` to a new line or is the only way to keep the previous functionality is to lock the version to 2.2?

**Pull request merged**
